addons/web/static/src/core/macro.js

```diff
@@ -0,0 +1,213 @@
+/** @odoo-module **/
+
+export const ACTION_HELPERS = {
+    click(el, _step) {
+        el.dispatchEvent(new MouseEvent("mouseover"));
+        el.dispatchEvent(new MouseEvent("mouseenter"));
+        el.dispatchEvent(new MouseEvent("mousedown"));
+        el.dispatchEvent(new MouseEvent("mouseup"));
+        el.click();
+        el.dispatchEvent(new MouseEvent("mouseout"));
+        el.dispatchEvent(new MouseEvent("mouseleave"));
+    },
+    text(el, step) {
+        // simulate an input (probably need to add keydown/keyup events)
+        this.click(el, step);
+        el.value = step.value;
+        el.dispatchEvent(new InputEvent("input", { bubbles: true }));
+        el.dispatchEvent(new InputEvent("change", { bubbles: true }));
+    },
+};
+
+class TimeoutError extends Error {}
+
+class Macro {
+    constructor(descr) {
+        this.name = descr.name || "anonymous";
+        this.timeoutDuration = descr.timeout || 0;
+        this.timeout = null;
+        this.currentIndex = 0;
+        this.interval = descr.interval ? Math.max(16, descr.interval) : 500;
+        this.isComplete = false;
+        this.steps = descr.steps;
+        this.onStep = descr.onStep || (() => {});
+        this.onError = descr.onError;
+        this.onTimeout = descr.onTimeout;
+        this.setTimer();
+    }
+
+    advance() {
+        if (this.isComplete) {
+            return;
+        }
+        const step = this.steps[this.currentIndex];
+        let trigger = step.trigger;
+        if (trigger) {
+            let el = null;
+            if (typeof trigger === "function") {
+                trigger = this.safeCall(trigger);
+                // so if it is some interesting value, it will be passed to
+                // the action function
+                el = trigger;
+            }
+            if (typeof trigger === "string") {
+                const $el = $(trigger);
+                el = $el.length && $el.eq(0).is(":visible:hasVisibility") ? $el[0] : null;
+            }
+            if (el) {
+                // we have a match!
+                this.advanceStep(el, step);
+            }
+        } else {
+            // a step without a trigger is just an action
+            this.advanceStep(null, step);
+        }
+    }
+
+    advanceStep(el, step) {
+        this.safeCall(this.onStep, el, step);
```

Comment:
if onStep crashes, we will still call the next action helper because it's not wrapped in a safeCall and safeCall catches the error without propagating it. Maybe handleError should rethrow the error by default?